src/engine/SumOfInfeasibilitiesManager.h

```diff
@@ -0,0 +1,193 @@
+
+    /* For debug use */
+    void setPhaseStatusInLastAcceptedPhasePattern( PiecewiseLinearConstraint
+                                                   *constraint,
+                                                   PhaseStatus phase );
+
+    void setPhaseStatusInCurrentPhasePattern( PiecewiseLinearConstraint
+                                              *constraint, PhaseStatus phase );
+
+
+private:
+    const List<PiecewiseLinearConstraint *> &_plConstraints;
+    // Used for the heuristic initialization of the phase pattern.
+    NLR::NetworkLevelReasoner *_networkLevelReasoner;
+    unsigned _numberOfVariables;
+    // Used for accessing the current variable assignment.
+    const ITableau &_tableau;
+
+    // Parameters that controls the local search heuristics
+    SoIInitializationStrategy _initializationStrategy;
+    SoISearchStrategy _searchStrategy;
+    double _probabilityDensityParameter;
+
+    /*
+      The representation of the current phase pattern (one linear phase of the
+      non-linear SoI function) as a mapping from PLConstraints to phase patterns.
+      We do not keep the concrete LinearExpression explicitly but will concretize
+      it on the fly. This makes it cheap to update the phase pattern.
+    */
+    Map<PiecewiseLinearConstraint *, PhaseStatus> _currentPhasePattern;
+
+    /*
+      The proposed update to the current phase pattern. For instance, it can
+      contain one of the ReLUConstraint in the _currentPhasePattern
+      with PhaseStatus flipped.
+    */
+    Map<PiecewiseLinearConstraint *, PhaseStatus> _lastAcceptedPhasePattern;
+
+    /*
+      The constraints in the current phase pattern (i.e., participating in the
+      SoI) stored in a Vector for ease of random access.
+    */
+    Vector<PiecewiseLinearConstraint *> _plConstraintsInCurrentPhasePattern;
+
+    /*
+      A local copy of the current variable assignment, which is refreshed via
+      the obtainCurrentAssignment() method.
+    */
+    Map<unsigned, double> _currentAssignment;
+
+    Statistics *_statistics;
+
+    /*
+      Clear _currentPhasePattern and _currentProposal
+    */
+    void resetPhasePattern();
+
+    /*
+      Set _currentPhasePattern according to the current input assignment.
+    */
+    void initializePhasePatternWithCurrentInputAssignment();
+
+    /*
+      Choose one piecewise linear constraint in the current phase pattern
+      and set it to a uniform-randomly chosen alternative phase status (for ReLU
+      this means we just flip the phase status).
+    */
+    void proposePhasePatternUpdateRandomly();
+
+    /*
+      Iterate over the piecewise linear constraints in the current phase pattern
+      to find one with the largest "reduced cost". See the "getReducedCost"
+      method below.
+      If no constraint has positive reduced cost (we are at a local optima), we
+      fall back to proposePhasePatternUpdateRandomly()
+    */
+    void proposePhasePatternUpdateWalksat();
+
+    /*
```

Comment:
"Reduced cost" is an overloaded term - the cost function manager also uses it. It it usually used to refer to the cost function in terms of the current non-basic variables. It sounds like it has a different meaning here; maybe change to costReduction or reductionInCost?

Reply:
Changed to cost reduction